lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -1127,6 +1127,78 @@ LogicalResult InferenceMapping::mapOperation(Operation *op) {
         }
       })
 
+      // Handle memories.
+      .Case<MemOp>([&](auto op) {
+        // Create constraint variables for all ports.
+        for (auto result : op.results())
+          declareVars(result, op.getLoc());
+
+        // This creates independent variables for every data port. Yet, what we
+        // actually want is for all data ports to share the same variable. To do
+        // this, we find the first data port declared, and use that port's vars
+        // for all the other ports.
+        LLVM_DEBUG(llvm::dbgs() << "Adjusting memory port variables:\n");
+        auto first = op.getResult(0);
+        auto firstType =
+            op.getPortType(0).getPassiveType().template cast<BundleType>();
+        auto firstKind = op.getPortKind(0);
+        unsigned firstBaseID;
+        switch (firstKind) {
+        case MemOp::PortKind::Read:
+          firstBaseID = firstType.getFieldID(3);
+          break;
+        case MemOp::PortKind::Write:
+          firstBaseID = firstType.getFieldID(3);
+          break;
+        case MemOp::PortKind::ReadWrite:
+          firstBaseID = firstType.getFieldID(4);
+          break;
+        }
+        unsigned firstMaxID = op.getDataType().getMaxFieldID();
+        LLVM_DEBUG(llvm::dbgs()
+                   << "- Data type max field ID: " << firstMaxID << "\n");
+        LLVM_DEBUG(llvm::dbgs() << "- First port type: " << firstType << "\n");
+        LLVM_DEBUG(llvm::dbgs()
+                   << "- First port base ID: " << firstBaseID << "\n");
```

Comment:
Can this all live in one `LLVM_DEBUG` scope?

Reply:
This comment reminds me, there is a debug print that I ruined with the aggregate types change, its printing `getFieldName(fieldRef)` or something.  When the type isn't from something with a name (i.e. a wire) its not printing anything useful, and we should probably make it print the IR Value as well.

Reply:
@seldridge Good idea!
@youngar Yeah that would be a great idea. How would we do that? The value names are discarded upon parsing, right?

Reply:
I'm not sure what the best thing to do is, but i think before it was just printing the IR value, so if we kept that maybe something like this?
```c++
#ifndef DDEBUG
  LLVM_DEBUG(llvm::dbgs() << "Expr " << *expr << " for " << fieldRef.getValue());
  if (fieldRef.getFieldID())
   LLVM_DEBUG(llvm::dbgs() << getFieldName(fieldRef));
 LLVM_DEBUG(llvm::dbgs() << "\n");
#endif
```

Reply:
Done!